app/src/main/java/com/humberto/tasky/agenda/data/task/TaskRepositoryImpl.kt

```diff
@@ -48,18 +56,31 @@ class TaskRepositoryImpl @Inject constructor(
         }
     }
 
-    override suspend fun deleteTask(taskId: String) {
-        val result = safeCall {
-            taskDao.deleteTask(taskId)
+    override suspend fun deleteTask(taskId: String): EmptyResult<DataError> {
+        taskDao.deleteTask(taskId)
+        return safeCall {
             agendaApi.deleteTask(taskId)
-        }
-        if(result is Result.Error) {
-            taskDao.insertModifiedTask(
-                ModifiedTaskEntity(
-                    taskId = taskId,
-                    modificationType = ModificationType.Deleted
+        }.onError { error ->
+            if (error.isRetryable()) {
+                taskDao.insertDeletedTaskSync(
+                    DeletedTaskSyncEntity(
+                        taskId = taskId,
+                        userId = localUserId!!
+                    )
                 )
-            )
+            }
+        }
+    }
+
+    override suspend fun syncPendingTasks() {
+        taskDao.getTasksPendingSync(
+            userId = localUserId ?: ""
+        ).map { taskPendingSyncEntity ->
+            safeCall {
+                agendaApi.createTask(taskPendingSyncEntity.task.toTask().toTaskRequest())
+            }.onSuccess {
+                taskDao.deleteTaskPendingSync(taskPendingSyncEntity.taskId)
+            }
```

Comment:
Here is another situation where you can perform a logout when the user ID is `null`. At this point, it makes sense to centralize the fetching of this user ID so that you can perform the check and perform logout (if needed) in one place.

`map` is used to transform a list, but here, you make a network call for each list item. Would it not make more sense to use `forEach` in that case? You also want the API calls to by asynchrounous since there is no need for one call to depend on the other.

To sync all the pending items (local changes that have not been sync with the remote storage yet), the following approach is great (it contains more than just Tasks, but the approach is the same 😄):

```
override suspend fun syncPendingAgendaItems() {
  withContext(dispatcherProvider.io) {

    // Step 1: Fetch the session ID of the logged in user to only sync their changes
    val userId = sessionStorage.getSession()?.userId ?: return@withContext

    // Step 2: For each type of sync operations, get all the pending transactions
    val createdReminders = async {
      pendingSyncDao.getAllPendingSyncReminderEntities(userId)
    }

    val deletedReminders = async {
      pendingSyncDao.getAllDeletedSyncReminderEntities(userId)
    }

    // ... continue here with updating reminders and then also handle tasks and events

    // Step 3: For each transaction, attempt the corresponding API call
    val createReminderJobs = createdReminders.await().map { pendingEntity ->
      launch {
        val reminder = pendingEntity.reminder.toAgendaItem()
        
        // When it fails, do nothing. When it succeeds, delete the pending operation from the local DB - we don't want to sync it again
        when (remoteDataSource.postReminder(reminder)) {
          is Result.Error -> Unit
          is Result.Success -> applicationScope.launch {
            pendingSyncDao.deleteReminderPendingSyncEntity(reminder.id)
          }.join()
        }
      }
    }

    val deletedReminderJobs = deletedReminders.await().map { pendingEntity ->
      launch {
        when (remoteDataSource.deleteAgendaItem(pendingEntity.reminderId)) {
          is Result.Error -> Unit
          is Result.Success -> applicationScope.launch {
            pendingSyncDao.deleteDeletedReminderSyncEntity(pendingEntity.reminderId)
          }.join()
        }
      }
    }

    // ... create jobs for the others as well

    // Step 4: Await all the operations in an async manner
    createReminderJobs.forEach { it.join() }
    deletedReminderJobs.forEach { it.join() }

    // ... call .join() for the other jobs as well
  }
}
```